sms-mms-import-to-signal.py

```diff
@@ -17,173 +35,229 @@ def get_contacts(cursor):
+            thread_id = rows[0]
+    return thread_id
 
 # parse the XML file generated by SMS Backup and Restore (by SyncTech)
-tree = ET.parse(sys.argv[1])
+logging.info('starting parsing xml file')
+tree = ET.parse(input)
 root = tree.getroot()
+logging.info('finished parsing xml file')
 
 # parse the sqlite database generated by github.com/bepaald/signalbackup-tools
-conn = sqlite3.connect(os.path.join(sys.argv[2], "database.sqlite"))
+conn = sqlite3.connect(os.path.join(output, "database.sqlite"))
 cursor = conn.cursor()
 
 smses = []
 mmses = []
 contacts_by_number = get_contacts(cursor)
 
 for r in root:
-
-    date_sent = r.attrib["date_sent"]
-    if date_sent == 0 or date_sent == "0" or date_sent == "" or not date_sent:
-        date_sent = r.attrib["date"]
-
+    date_sent = r.attrib.get("date_sent","")
+    if date_sent in [0, "0", ""] or len(date_sent) < 13:
+        date_sent = r.attrib.get("date","")
+    addrs = get_addrs(r)
     address = False
-    add = r.attrib["address"].replace("-", "")
-
-    # some sms conversations seem to have multiple recipients separated by a tilda
-    # I don't really know how to craft a signal group chat (although I do see a group table)
-    # So we're just going to grab one of the contacts in the group and make them the
-    # recipient ... sure wish the Signal database had some documentation ...
+    add = r.attrib["address"].replace("-", "").replace("+61","0")
+    add_list = []
     if '~' in add:
         for a in sorted(add.split("~")):
             try:
                 address = contacts_by_number[a]
-                print("Coerced group of ", add, " to a single contact: ", a)
             except KeyError:
-                pass
-
+                address = add_recipient(a, cursor)
+            finally:
+                add_list.append(address)
+        else:
+            for addr in addrs:
+                if addr.items()[1][1] == '137':
+                    address = contacts_by_number[addr.items()[0][1].replace("-", "").replace("+61","0")]
+        add_list = [*set(add_list)]
     if not address:
         try:
             address = contacts_by_number[add]
         except KeyError:
-            pass
-
-    if not address:
-        print("\nSkipping ", r.tag, " because couldn't link address to a contact: ", r.attrib["address"])
-        print(r.items())
-        # sys.exit(0)
-        # cursor.execute("insert into recipient (phone, system_display_name) values (?, ?)", (r.attrib["address"],.....))
-        continue
-
+            address = add_recipient(add, cursor)
+    row = {}
     if r.tag == "sms":
-
-        # magic Signal numbers
-        typ = 10485783 if str(r.attrib["type"]) == "2" else 10485780
-
-        row = (
-            address,
-            r.attrib["date"],
-            date_sent,
-            1,  # "read"
-            -1,  # "status"
-            int(typ),
-            r.attrib["body"],
-        )
+        #sms sent is type 87, sms received is type 20
```

Comment:
it's interesting that your magic numbers are different to my magic numbers... wonder what's happening there. Is it because you've managed to figure out how to insert the messages are "sms" flavoured messages instead of as I was doing, sort of faking them as pretend signal pseudo messages...

Reply:
that is exactly correct, these magic numbers are to make them appear as SMS/MMS rather than signal messages

Reply:
Given Signal's (unbelievable) announcement yesterday about dropping SMS, and indeed their suggestion that we'll need to export all SMS out of signal, because they'll be deleted otherwise - we might need to revisit these magic numbers.

Eg, I think I'm going to fork this project into a very similar one, that'll crack open a Signal backup and update all the message types - cloak the SMSes as Signal messages so they won't be deleted. Do you reckon that'll work?

Starting to feel a bit miffed about this whole thing :(

Reply:
Now it sounds like they're not going to delete the messages:

https://community.signalusers.org/t/signal-blog-removing-sms-support-from-signal-android-soon/47954/18

Btw was I doing something wrong with the MMSes? I only updated the `sms` table, I wasn't sure what needed to be updated in the `mms` table.

Reply:
you're right, according to that it looks like they're not going to be deleting SMS/MMS from the Signal database, just cutting off access to send/receive.
https://www.signal.org/blog/sms-removal-android/
this indicates the opposite, however - so there's potentially still value in your fork to mark things as Signal messages just in case, until we know for sure.

with the MMS table, my memory tells me there was a different message type/message box value for Signal picture messages - let me have a look-see at a backup now.

Reply:
so, looks like the `m_type ` for the `mms` table stays the same between Signal and 'regular' MMS messages.
the `msg_box `becomes 10485780 where it was 20, and 10485783 where it was 87, eg
`update mms
set msg_box = case msg_box when 20 then 10485780 when 87 then 10485783 end
where msg_box in (20,87);`